app-ng/src/job.lisp

```diff
@@ -0,0 +1,113 @@
+(in-package #:qvm-app-ng)
+
+(deftype job-status ()
+  '(member fraiche running finished interrupted error))
+
+(defstruct (job (:constructor %make-job))
+  "A job is an data-structure that permits a given function to be run in a threaded (and thread-safe) environment."
+  lock
+  threader
+  thread
+  %result                              ; I define this accessor below.
+  (status 'fraiche :type job-status)
+  error)
+
+(defmacro with-job-lock ((job &key (wait t)) &body body)
+  "Evaluate BODY with the LOCK acquired for the duration of BODY. If WAIT is T, block until the lock is available."
+  `(progn
+     (when (bt:acquire-lock (job-lock ,job) ,wait)
+       (unwind-protect
+           (progn ,@body)
+         (bt:release-lock (job-lock ,job))))))
+
+(defmethod print-object ((job job) stream)
+  (print-unreadable-object (job stream :type nil :identity t)
+    (format stream "QVM Job (status ~A)" (job-status job))))
+
+;; The %x-unsafe variants below do not acquire a lock.
+
+(defun %job-running-p-unsafe (job)
+  (eq (job-status job) 'running))
+
+(defun job-running-p (job)
+  "Is the JOB in the RUNNING state."
+  (with-job-lock (job)
+    (%job-running-p-unsafe job)))
+
+(defun %job-interrupted-p-unsafe (job)
+  (eq (job-status job) 'interrupted))
+
+(defun job-interrupted-p (job)
+  "Is the JOB in the INTERRUPTED state?"
+  (with-job-lock (job)
+    (%job-interrupted-p-unsafe job)))
+
+(defun %job-finished-p-unsafe (job)
+  (eq (job-status job) 'finished))
+
+(defun job-finished-p (job)
+  "Is the JOB in the FINISHED state?"
+  (with-job-lock (job)
+    (%job-finished-p-unsafe job)))
+
+(defun make-job (fn)
+  "Make a JOB object which will, under JOB-START, run FN in a thread.
+
+The returned JOB can be repeatedly started if the JOB-STATUS is FINISHED (or INTERRUPTED). Only interact with the job after acquiring the lock (or alternatively by using WITH-JOB-LOCK)."
+  (let* ((job (%make-job))
+         (thr-fn (lambda ()
```

Comment:
I would initialize most slots of the job _before_ the thread is spawned, since you might have a race condition on accessing the object